File: Object/Brdf/Phong.cpp
#define _USE_MATH_DEFINES
#include "Object/Brdf/Phong.hpp"

#include "Math/Normal.hpp"
#include "Math/Vector.hpp"
#include "Math/OrthonormalBasis.hpp"

#include <cmath>
#include <algorithm>

namespace Object::Brdf {
    Phong::Phong(float strength, float power)
    {
        mStrength = strength;
        mPower = power;
    }

    Math::Color Phong::reflected(const Math::Vector &dirIn, const Math::Normal &nrm, const Math::Vector &dirOut, const Math::Color &) const
    {
        Math::Vector dirReflect = -(dirIn - Math::Vector(nrm) * (2 * (nrm * dirIn)));

        float dot = dirReflect * dirOut;
        float coeff = 0;
        if(dot > 0) {
            coeff = std::pow(dot, mPower);
        }

        return Math::Color(1, 1, 1) * mStrength * coeff * (mPower + 1) / (2 * (float)M_PI);
    }

    Math::Color Phong::transmitted(const Math::Vector &, const Math::Normal &, const Math::Color &) const
    {
        return Math::Color(1, 1, 1) * (1.0f - mStrength);
    }

    Math::Vector Phong::sample(Math::Sampler::Base &sampler, const Math::Normal &nrm, const Math::Vector &dirOut) const
    {
        Math::Point2D samplePoint = sampler.getValue2D();
        float phi = 2 * M_PI * samplePoint.u();
        float theta = std::acos(std::pow(samplePoint.v(), 1.0f / (mPower + 1)));

        Math::OrthonormalBasis basis(dirOut);

        Math::Vector dirReflect = basis.localToWorld(Math::Vector::fromPolar(phi, M_PI / 2 - theta, 1));
        Math::Vector dirIn = -(dirReflect - Math::Vector(nrm) * (dirReflect * nrm * 2));

        return dirIn;
    }

    float Phong::pdf(const Math::Vector &dirIn, const Math::Normal &nrm, const Math::Vector &dirOut) const
    {
        float coeff = 0;
        Math::Vector dirReflect = -(dirIn - Math::Vector(nrm) * (dirIn * nrm * 2));
        float dot = dirReflect * dirOut;
        if (dot > 0) {
            coeff = std::pow(dot, mPower);
        }

        float pdf = coeff * (mPower + 1) / (2 * M_PI);
        return std::min(pdf, 1000.0f);
    }

    void Phong::writeProxy(BrdfProxy &proxy) const
    {
        proxy.type = BrdfProxy::Type::Phong;
        proxy.phong.strength = mStrength;
        proxy.phong.power = mPower;
    }
}

File: src/object/brdf/phong.rs
use crate::object;
use crate::geo;

use geo::Vec3;
use geo::Normal3;
use geo::OrthonormalBasis;

use object::Color;
use object::Sampler;

use std::f32::consts::PI;

pub struct Phong {
    strength : f32,
    power : f32
}

impl Phong {
    pub fn new(strength : f32, power : f32) -> Phong {
        Phong {strength, power}
    }
}

impl object::Brdf for Phong {
    fn reflected(&self, dir_in : Vec3, nrm : Normal3, dir_out : Vec3, _albedo : Color) -> Color {
        let dir_reflect = -(dir_in - nrm.to_vec3() * (2.0 * nrm * dir_in));
        let dot = dir_reflect * dir_out;
        let coeff = if dot > 0.0 { dot.powf(self.power) } else { 0.0 };
        
        return Color::ONE * self.strength * coeff * (self.power + 1.0) / (2.0 * PI);
    }

    fn transmitted(&self, _dir_in : Vec3, _normal : Normal3, _albedo : Color) -> Color {
        return Color::ONE * (1.0 - self.strength);
    }

    fn lambert(&self) -> f32 {
        return self.strength;
    }

    fn sample(&self, sampler : &mut dyn Sampler, nrm : Normal3, dir_out : Vec3) -> Vec3
    {
        let sample_point = sampler.get_value2();
        let phi = 2.0 * PI * sample_point.u;
        let theta = sample_point.v.powf(1.0 / (self.power + 1.0)).acos();
        
        let basis = OrthonormalBasis::new(dir_out);

        let dir_reflect = basis.local_to_world(Vec3::from_polar(phi, PI / 2.0 - theta, 1.0));
        let dir_in = -(dir_reflect - nrm.to_vec3() * (dir_reflect * nrm * 2.0));
        return dir_in;
    }

    fn pdf(&self, dir_in : Vec3, nrm : Normal3, dir_out : Vec3) -> f32 {
        let mut coeff = 0.0;
        let dir_reflect = -(dir_in - nrm.to_vec3() * (dir_in * nrm * 2.0));
        let dot = dir_reflect * dir_out;
        if dot > 0.0 {
            coeff = dot.powf(self.power);
        }

        let pdf = coeff * (self.power + 1.0) / (2.0 * PI);
        return pdf.min(1000.0);
    }
}